### PR TITLE
Fix length reported by DALI FW iterators when DROP policy is used

### PR DESCRIPTION
**Pull request by JanuszL**
- DALI reported the wrong length of the FW iterator when the drop policy was used. This PR corrects this.

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It fixes the length reported by DALI FW iterators when DROP policy is used

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     DALI reported the wrong length of the FW iterator when the drop policy was used. Fix the calculation logic
 - Affected modules and functionalities:
     base_iterator
 - Key points relevant for the review:
     NA
 - Validation and testing:
     Test is added
 - Documentation (including examples):
     NA

Relates to https://github.com/NVIDIA/DALI/issues/2607
**JIRA TASK**: *[NA]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/test/python/test_fw_iterators.py [hunk 1]
```diff
@@ -339,13 +339,17 @@ def check_mxnet_iterator_pass_reader_name(shards_num, pipes_number, batch_size,
...
         for it in iter(dali_train_iter):
             for id in range(pipes_number):
                 tmp = it[id].data[0].squeeze(-1).asnumpy().copy()
                 if it[id].pad:
                     tmp = tmp[0:-it[id].pad]
                 img_ids_list[id].append(tmp)
             sample_counter += batch_size
+            length -= 1
+
+        assert length == 0, "Iterator has reported wrong size {}, while {} iteration of data has been provide".format(orig_length, orig_length - length)
```
Comment (banasraf) [inline dali/test/python/test_fw_iterators.py:352]:
```suggestion
        assert length == 0, "The iterator has reported the length of {} but provided {} iterations (//or elements).".format(orig_length, orig_length - length)
```
??

**Review by banasraf — APPROVED** [review]

File: dali/test/python/test_fw_iterators.py (hunk 1 above)
Comment (JanuszL) [reply dali/test/python/test_fw_iterators.py:352]: Fixed

File: dali/test/python/test_fw_iterators.py [hunk 2]
```diff
@@ -339,13 +339,17 @@ def check_mxnet_iterator_pass_reader_name(shards_num, pipes_number, batch_size,
...
         for it in iter(dali_train_iter):
             for id in range(pipes_number):
                 tmp = it[id].data[0].squeeze(-1).asnumpy().copy()
                 if it[id].pad:
                     tmp = tmp[0:-it[id].pad]
                 img_ids_list[id].append(tmp)
             sample_counter += batch_size
+            length -= 1
+
+        assert length == 0, "The iterator has reported the length of {} but provided {} iterations.".format(orig_length, orig_length - length)
```
Comment (jantonguirao) [inline dali/test/python/test_fw_iterators.py:352]: just FYI, you can also do:
```
 assert length == 0, f"The iterator has reported the length of {orig_length} but provided {orig_length - length} iterations."
 ```

**Review by jantonguirao — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

File: dali/test/python/test_fw_iterators.py (hunk 2 above)
Comment (JanuszL) [reply dali/test/python/test_fw_iterators.py]: Done

**Pull request merged**